src/scheduler/worker_monitor.rs

```diff
@@ -0,0 +1,238 @@
+//! This module contains `WorkerMonitor` and related types.  It purposes includes:
+//!
+//! -   allowing workers to park,
+//! -   letting the last parked worker take action, and
+//! -   letting workers and mutators to notify workers when workers are given things to do.
+
+use std::sync::{Condvar, Mutex};
+
+use crate::vm::VMBinding;
+
+use super::{
+    worker_goals::{WorkerGoal, WorkerGoals},
+    GCWorker,
+};
+
+/// The result type of the `on_last_parked` call-back in `WorkMonitor::park_and_wait`.
+/// It decides how many workers should wake up after `on_last_parked`.
+pub(crate) enum LastParkedResult {
+    /// The last parked worker should wait, too, until more work packets are added.
+    ParkSelf,
+    /// The last parked worker should unpark and find work packet to do.
+    WakeSelf,
+    /// Wake up all parked GC workers.
+    WakeAll,
+}
+
+/// A data structure for synchronizing workers with each other and with mutators.
+///
+/// Unlike `GCWorkerShared`, there is only one instance of `WorkerMonitor`.
+///
+/// -   It allows workers to park and unpark.
+/// -   It allows mutators to notify workers to schedule a GC.
+pub(crate) struct WorkerMonitor {
+    /// The synchronized part.
+    sync: Mutex<WorkerMonitorSync>,
+    /// Workers wait on this when idle.  Notified if workers have things to do.  That include:
+    /// -   any work packets available, and
+    /// -   any field in `sync.goals.requests` set to true.
+    workers_have_anything_to_do: Condvar,
```

Comment:
Not sure why you changed the name of the `Condvar`.

Reply:
The name `work_available` may be ambiguous.  `work` may be interpreted as "work packet", but now the Condvar may also be notified if a "goal" is set.